### PR TITLE
a8n: Add checkState to ExternalChangeset resolver

### PR DESCRIPTION
**Pull request by ryanslade**
This change adds support for loading github pull request commits and their associated statuses from github.

In order to keep the PR small a follow up PR's will:
* Add support for webhooks
* Add bitbucket support

### CODE REVIEW

**Comment by eseliger** [comment]
@ryanslade I added some initial UI here, so you can also use the UI to test it if you like! 

do we plan on updating the UI regularly to get notified of new statuses? 🤔
ways to implement: 
- polling
- out of scope just to mention: somewhat real-time updates
- don't account for it for now

**Comment by ryanslade** [comment]
> @ryanslade I added some initial UI here, so you can also use the UI to test it if you like!
> 
> do we plan on updating the UI regularly to get notified of new statuses? 🤔
> ways to implement:
> 
> * polling
> * out of scope just to mention: somewhat real-time updates
> * don't account for it for now

I think for now polling is ok. Ideally long term we probably want it to update in real time similar to GitHub but that would require a lot of changes on both sides.

**Comment by ryanslade** [comment]
@eseliger It looks like it's quite common for the status to be empty. I think we may want an "Unknown" option on our end too, what do you think?

**Comment by eseliger** [comment]
> I think for now polling is ok. Ideally long term we probably want it to update in real time similar to GitHub but that would require a lot of changes on both sides.

I agree

> @eseliger It looks like it's quite common for the status to be empty. I think we may want an "Unknown" option on our end too, what do you think?

I was going to account for that with the Status being optional - but we can make it more explicit, too! If so, I would add a visual representation of the new state also

**Comment by ryanslade** [comment]
@eseliger Ok, don't worry. I'll return a `nil` status in that case

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -764,6 +771,10 @@ type ExternalChangeset implements Node {
     # The diff of this changeset.
     # Only returned if the changeset has not been merged or closed.
     diff: RepositoryComparison
+
+    # The state of the continuous integration checks on this changeset
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
    # The state of the continuous integration checks on this changeset.
```

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -764,6 +771,10 @@ type ExternalChangeset implements Node {
...
+    # It can be null if no checks have been configured
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
    # It can be null if no checks have been configured.
```

File: enterprise/internal/a8n/resolvers/changesets.go [hunk 3]
```diff
@@ -205,14 +205,31 @@ func (r *changesetResolver) ReviewState(ctx context.Context) (a8n.ChangesetRevie
...
-	events := make(a8n.ChangesetEvents, len(es))
-	for i, e := range es {
-		events[i] = e
+	events := a8n.ChangesetEvents(es)
+	sort.Sort(events)
+	return events.ReviewState()
+}
+
+func (r *changesetResolver) CheckState(ctx context.Context) (*a8n.ChangesetCheckState, error) {
+	// ChangesetEvents are currently only implemented for GitHub. For other
```
Comment (mrnugget) [inline enterprise/internal/a8n/resolvers/changesets.go]: This comment is wrong

File: enterprise/internal/a8n/resolvers/changesets.go [hunk 4]
```diff
@@ -205,14 +205,31 @@ func (r *changesetResolver) ReviewState(ctx context.Context) (a8n.ChangesetRevie
...
+		// TODO: Implement for BitBucket
+		return nil, nil
 	}
 
-	sort.Sort(events)
+	opts := ee.ListChangesetEventsOpts{
+		ChangesetIDs: []int64{r.Changeset.ID},
+		Limit:        -1,
+	}
+	es, _, err := r.store.ListChangesetEvents(ctx, opts)
```
Comment (mrnugget) [inline enterprise/internal/a8n/resolvers/changesets.go]: I think we should wrap the loading of the changeset events in a `sync.Once` (see the other resolvers that have `compute` method) so that we don't load the complete list of changeset events twice (see `ReviewState` above)

File: internal/a8n/types.go [hunk 5]
```diff
@@ -541,6 +551,40 @@ func (ce ChangesetEvents) ReviewState() (ChangesetReviewState, error) {
...
 }
 
+func (ce ChangesetEvents) CheckState() (*ChangesetCheckState, error) {
+	commits := make([]github.Commit, 0, len(ce))
+	for _, e := range ce {
+		if e.Kind != ChangesetEventKindGitHubCommit {
+			continue
+		}
+		commits = append(commits, e.Metadata.(*github.PullRequestCommit).Commit)
+	}
```
Comment (mrnugget) [inline internal/a8n/types.go]: I wonder: can't you get rid of the `sort.Slice` and the aggregation of all `commits` by remembering the index and commitdate of the newest commit? Something like:

```go
	var (
		newestCommitDate time.Time
		newestIndex int
	)
	for i, e := range ce {
		if e.Kind != ChangesetEventKindGitHubCommit {
			continue
		}
		commit := e.Metadata.(*github.PullRequestCommit).Commit
		if commit.CommitDate.After(newestCommitDate) {
			newestCommitDate = commit.CommitDate
			newestCommitIndex = i
		}
	}

	latest := ce[newestIndex].(*github.PullRequestCommit).Commit
```

File: internal/extsvc/github/pulls.go [hunk 6]
```diff
@@ -623,137 +622,213 @@ func (c *Client) GetOpenPullRequestByRefs(ctx context.Context, owner, name, base
...
+        actor {
+          ...actor
+        }
+        assignee {
+          ...actor
+        }
+        createdAt
+      }
+    }
+  }
```
Comment (mrnugget) [inline internal/extsvc/github/pulls.go]: Just a note: the reformatting makes it hard to tell what was changed. I know that it's not always feasible or practical to do so, but I think saving the reformatting for a separate PR helps reviewers :)

Comment (ryanslade) [reply internal/extsvc/github/pulls.go]: Yeah, sorry about that. I used the "pretty" output from the github explorer and so I'm hoping from now on we can use that to avoid having to format manually. I'll add a comment above the fragment

File: enterprise/internal/a8n/resolvers/changesets.go (hunk 3 above)
Comment (ryanslade) [reply enterprise/internal/a8n/resolvers/changesets.go]: Ah, it was copied from ReviewState above as I used that as a template. I'll update both.

**Comment by ryanslade** [comment]
@mrnugget PTAL

**Comment by ryanslade** [comment]
So, slight complication.

The code so far is using GitHib commit status which is the "old" way of doing things but works with, for example, `sourcegraph/sourcegraph`. However, I've just tried testing with github actions and that uses `CheckRuns` and `CheckSuites` instead. I think in the long term we're going to need to support both. However, at the moment the Checks spec is still in beta:
https://developer.github.com/v3/checks/runs/#create-a-check-run

I'll create a new issue to specifically target the new checks api and come back to that later in the iteration if I have time

**Comment by felixfbecker** [comment]
Oh yeah, forgot that there is also the conflict between "checks" and "commit status". I actually thought checks are only available to GitHub apps?

**Comment by ryanslade** [comment]
> Oh yeah, forgot that there is also the conflict between "checks" and "commit status". I actually thought checks are only available to GitHub apps?

I'm working on the webhook integration and when testing with GitHub actions it sends me `CheckRun` and `CheckSuite` events

**Comment by ryanslade** [comment]
Heads up that now that I've started looking into webhooks it's going to change the way I need to implement some of the existing code so I'm going to continue working on this branch.

I'm going to close this now so that nobody is spammed while I continue working and I'll reopen when it's ready for review

**Comment by ryanslade** [comment]
@eseliger Can you have a look at the spacing and icons for the different check statuses?

image

It would be great if they were more consistent, especially the small green dot for `PASSED`. Also, at least to me, I think there needs to be more spacing between the icons and the boxes showing code additions

**Comment by eseliger** [comment]
How about that?
image

**Comment by ryanslade** [comment]
> How about that?
> image

Nice! I like it 👍

**Comment by mrnugget** [comment]
I'm concerned that it's not immediately obvious what these icons signal. Is there a visual way to make clear that these states are _not_ the changeset state ("publishing", "created", "errored") but CI checks? Can we maybe add a label next to them? (At the very least a hover tooltip).

Since we also want to improve how we display a changeset's state (draft -> publishing -> created -> errored), maybe adding visual elements for both goes hand in hand here?

**Comment by felixfbecker** [comment]
I think we should use equivalent icons to GitHub. The yellow dot is smaller, and the others are not in a dot. (Refined?) GitHub puts them after the title, I find that quite intuitive ("this is the change, and it was checked that it works correctly")

image

**Comment by mrnugget** [comment]
> I think we should use equivalent icons to GitHub.

Yeah, I like GitHub's icons here. But we also need to consider that our use case is _slightly_ different:

1. GitHub only needs to display one state. We want to display two (state of changeset creation on codehost, state of checks)
2. GitHub's title links are immediately obvious as such. @unknwon reported to me that he didn't immediately recognize how to go from the changeset entry to the pull request on GitHub, which made me think that we should add an `ExternalLinkIcon` next to the title (along with adjusting colors etc.). With such an icon: does the title line get too busy?

**Comment by felixfbecker** [comment]
> GitHub's title links are immediately obvious as such. @unknwon reported to me that he didn't immediately recognize how to go from the changeset entry to the pull request on GitHub, which made me think that we should add an ExternalLinkIcon next to the title (along with adjusting colors etc.). 

I think this is a larger question, because this is just the standard link color we use everywhere in the app. If that is not obvious to be clickable, it would mean we would have to revamp it everywhere, because it wouldn't make sense to add icons as a workaround in _some_ places if the link color is generally not obviously clickable. Interestingly though, GitHub's PR titles are actually black (and only the title is clickable). I'd be curious to hear if you feel the same way about those @unknwon, or whether they it's that they give an additional hint that we could add too (hover background shade? Blue hover color?).

TL;DR I think we should discuss that separately but happy to have that discussion. Moving the checks somewhere else if we take such action would not be hard.

**Comment by mrnugget** [comment]
> I think this is a larger question, because this is just the standard link color we use everywhere in the app. If that is not obvious to be clickable, it would mean we would have to revamp it everywhere, because it wouldn't make sense to add icons as a workaround in _some_ places if the link color is generally not obviously clickable.

I don't think we need to "revamp it everywhere" to make this particular external link stand out a little more 🙂 Context always plays a role and just because something doesn't work in one context doesn't mean it it won't work in another one.

In this case here, for example, the link _right next to it_, the repository name, is a light grey:

This is the title of the campaign - Sourcegraph-2020-01-30-11-21-13

That doesn't help distinguishing links from non-links, but it makes sense for them to be different, I guess, because the link to the changeset on the codehost is an _external link_ and those sometimes have this icon next to them (see "All keywords..." at the bottom):

This is the title of the campaign - Sourcegraph-2020-01-30-11-16-51

(Note: that link is also not blue)

> TL;DR I think we should discuss that separately but happy to have that discussion. Moving the checks somewhere else if we take such action would not be hard.

Yep, agree, as I said, I probably should open a separate issue for this.

**Comment by unknwon** [comment]
> I think this is a larger question, because this is just the standard link color we use everywhere in the app. If that is not obvious to be clickable, it would mean we would have to revamp it everywhere, because it wouldn't make sense to add icons as a workaround in _some_ places if the link color is generally not obviously clickable. Interestingly though, GitHub's PR titles are actually black (and only the title is clickable). I'd be curious to hear if you feel the same way about those @unknwon, or whether they it's that they give an additional hint that we could add too (hover background shade? Blue hover color?).

I think I wasn't crystal clear when I gave the feedback to @mrnugget yesterday, the problem/confusion I encountered was:
- I know the title looks like a clickable link
- I also see same GitHub PR icon (was thinking, might be clickable?)
- And the GitHub repository URL (feels like a clickable thing with first intuition)

Thus I start wondering which one is the "correct" think to click could lead me to the GitHub PR page. For me, I don't think it's about colors, it's that there is no clear indication which one would lead me to the code host.

One easy solution to me is to do the same thing as we do in the file view:
image

A GitHub Logo is a clear indication it would take me somewhere related to GitHub, and a hover tip explains in detail, could just say "See pull request on GitHub/Bitbucket Server", etc.

**Comment by ryanslade** [comment]
I've reopened this after merging from master.

I'd like to get this merged in so that it doesn't get too stale and I'll do follow up PR's to implement:

* CheckRun and CheckSuite for GitHub
* BitBucket support
* Support webhooks for GitHub
* Support webhooks for BitBucket

I don't believe any of the above will require frontend changes

**Comment by ryanslade** [comment]
I'm going to merge this. We can tweak the UI in followup PR's along with the outstanding backend work
